prime-router/src/main/kotlin/azure/WorkflowEngine.kt

```diff
@@ -288,6 +304,161 @@ class WorkflowEngine(
+        when {
+            options == Options.SkipSend -> {
+                // Note that SkipSend should really be called SkipBothTimingAndSend  ;)
+                val event = ReportEvent(Event.EventAction.NONE, report.id)
+                this.dispatchReport(event, report, actionHistory, receiver, txn, context)
+                loggerMsg = "Queue: ${event.toQueueMessage()}"
+            }
+            receiver.timing != null && options != Options.SendImmediately -> {
+                val time = receiver.timing.nextTime()
+                // Always force a batched report to be saved in our INTERNAL format
+                val batchReport = report.copy(bodyFormat = Report.Format.INTERNAL)
+                val event = ReceiverEvent(Event.EventAction.BATCH, receiver.fullName, time)
+                this.dispatchReport(event, batchReport, actionHistory, receiver, txn, context)
+                loggerMsg = "Queue: ${event.toQueueMessage()}"
+            }
+            receiver.format == Report.Format.HL7 -> {
+                report
+                    .split()
+                    .forEach {
+                        val event = ReportEvent(Event.EventAction.SEND, it.id)
+                        this.dispatchReport(event, it, actionHistory, receiver, txn, context)
+                    }
+                loggerMsg = "Queued to send immediately: HL7 split into ${report.itemCount} individual reports"
+            }
+            else -> {
+                val event = ReportEvent(Event.EventAction.SEND, report.id)
+                this.dispatchReport(event, report, actionHistory, receiver, txn, context)
+                loggerMsg = "Queued to send immediately: ${event.toQueueMessage()}"
+            }
+        }
+        context.logger.info(loggerMsg)
+    }
+
+    /**
+     * Handle a receiver specific event. Fetch all pending tasks for the specified receiver and nextAction
+     *
+     * @param messageEvent that was received
+     * @param context execution context
+     * @param actionHistory action history being passed through for this message process
+     */
+
+    fun handleProcessEvent(
+        messageEvent: ProcessEvent,
+        context: ExecutionContext,
+        actionHistory: ActionHistory
+    ) {
+        val errors: MutableList<ResultDetail> = mutableListOf()
+        val warnings: MutableList<ResultDetail> = mutableListOf()
+
+        db.transact { txn ->
+            val task = db.fetchAndLockTask(messageEvent.reportId, txn)
+            val id = task.reportId
+            val reportFile = db.fetchReportFile(id, org = null, txn)
+            val header = createHeader(task, reportFile, null, null, null)
+            val currentAction = Event.EventAction.parseQueueMessage(task.nextAction.literal)
+
+            // get sender record
+            val sender = settings.findSender(reportFile.sendingOrg + "." + reportFile.sendingOrgClient)
+
+            //  create report
+            val report = createReport(
+                sender!!,
```

Comment:
Lots of !! here.  Shouldn't you be checking this variables before using this?

Reply:
There is very low chance of a null sender - this has already been pulled as part of initial receive. Updated to check anyway just in case something odd happens.